### PR TITLE
Use walkdir based copy + don't copy .git, fix "Permission denied" build errors

### PR DESCRIPTION
**Pull request by cjdelisle**
Switch from using xcopy/cp to using walkdir + fs::copy to improve reliability, also don't copy .git directory because read-only files cause failed rebuilds

Fixes errors such as:
```
  cp: cannot create regular file '/Users/user/wrk/play/electrum/contrib/packetcrypt_rs/packetcrypt-dll/target/release/build/libsodium-sys-6b0dfc7ac10b8aa6/out/source/libsodium/./.git/objects/pack/pack-762
2e108e422314744293f1577160ccc4aced3f6.pack': Permission denied                                                                                                                                              
  cp: cannot create regular file '/Users/user/wrk/play/electrum/contrib/packetcrypt_rs/packetcrypt-dll/target/release/build/libsodium-sys-6b0dfc7ac10b8aa6/out/source/libsodium/./.git/objects/pack/pack-762
2e108e422314744293f1577160ccc4aced3f6.idx': Permission denied                                                                                                                                               
  thread 'main' panicked at 'Failed to copy sources into build directory: exit code: 1', /Users/user/.cargo/git/checkouts/sodiumoxide-9b8c3ad42446e2d9/c1581f0/libsodium-sys/build.rs:370:13
  note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
warning: build failed, waiting for other jobs to finish...
error: build failed
```

### CODE REVIEW

File: libsodium-sys/build.rs [hunk 1]
```diff
@@ -346,29 +347,22 @@ fn build_libsodium() {
...
-
-    match cp_status {
-        Ok(status) if status.success() => (),
-        Ok(status) => {
-            panic!("Failed to copy sources into build directory: {}", status);
+    // Skip .git because it's marked read-only and that causes problems re-building
+    for entry in walkdir::WalkDir::new("libsodium")
+        .into_iter()
+        .filter_entry(|e| e.file_name().to_str().map(|s| s != ".git").unwrap_or(false))
+        .filter_map(|me| if let Ok(e) = me { Some(e) } else { None })
```
Comment (adamreichold) [inline libsodium-sys/build.rs:354]: Isn't this `filter_map(Result::ok)`?

File: libsodium-sys/build.rs [hunk 2]
```diff
@@ -346,29 +347,22 @@ fn build_libsodium() {
...
-    };
-
-    match cp_status {
-        Ok(status) if status.success() => (),
-        Ok(status) => {
-            panic!("Failed to copy sources into build directory: {}", status);
+    // Skip .git because it's marked read-only and that causes problems re-building
+    for entry in walkdir::WalkDir::new("libsodium")
+        .into_iter()
+        .filter_entry(|e| e.file_name().to_str().map(|s| s != ".git").unwrap_or(false))
```
Comment (adamreichold) [inline libsodium-sys/build.rs:353]: Maybe that comparison could go the other way around, e.g. something like
```
e.file_name() == Some(OsStr::new(".git"))
```

**Comment by cjdelisle** [comment]
@adamreichold Thanks for the comments, implemented both.

File: libsodium-sys/build.rs [hunk 3]
```diff
@@ -346,29 +347,22 @@ fn build_libsodium() {
...
-    };
-
-    match cp_status {
-        Ok(status) if status.success() => (),
-        Ok(status) => {
-            panic!("Failed to copy sources into build directory: {}", status);
+    // Skip .git because it's marked read-only and that causes problems re-building
+    for entry in walkdir::WalkDir::new("libsodium")
+        .into_iter()
+        .filter_entry(|e| e.file_name() == OsStr::new(".git"))
```
Comment (kpcyrd) [inline libsodium-sys/build.rs:353]:
```suggestion
        .filter_entry(|e| e.file_name() != OsStr::new(".git"))
```
This needs to be flipped. To reproduce the error try to delete `target/` and rebuild.

File: libsodium-sys/build.rs [hunk 4]
```diff
@@ -346,29 +347,22 @@ fn build_libsodium() {
...
+        .filter_map(Result::ok)
+    {
+        let outpath = out_dir.join("source").join(entry.path());
+        //println!("{:?}", &outpath);
```
Comment (kpcyrd) [inline libsodium-sys/build.rs:357]:
```suggestion
```
This should be deleted.

**Review by kpcyrd — CHANGES_REQUESTED** [review]

File: libsodium-sys/build.rs (hunk 3 above)
Comment (cjdelisle) [reply libsodium-sys/build.rs]: Whoops, that was a bit silly. Pushed an update, thanks.

**Review by kpcyrd — APPROVED** [review]

**Comment by Dylan-DPC** [comment]
testing new config
@bors try

**Comment by Dylan-DPC** [comment]
bors: try

**Comment by kpcyrd** [comment]
@cjdelisle we've fixed the ci, could you rebase your branch to get rid of the merge commits? :)

**Comment by cjdelisle** [comment]
@kpcyrd done but it looks like your CI thing still doesn't work

**Review by kpp — APPROVED** [review]

**Comment by kpp** [comment]
bors: r+

**Pull request merged**